File: Dynamic Programming/edit_distance.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/timeb.h>
#include <string.h>
int min(int a, int b, int c)
{
		if (a <= b && a <= c)
				return(a);
		else if (b <= c && b <= a)
				return(b);
		else
				return(c);
}


int editDistanceRecursive(char *arra, char *arrb, int asize, int bsize)
{
		if (asize == 0)
				return(bsize);
		if (bsize == 0)
				return(asize);
		int tmp = 1;
		if (arra[asize - 1] == arrb[bsize - 1])
				tmp = 0;
		return(min(editDistanceRecursive(arra, arrb, asize-1, bsize-1) + tmp,
				   editDistanceRecursive(arra, arrb, asize-1, bsize) + 1,
				   editDistanceRecursive(arra, arrb, asize, bsize-1) + 1));
}


int editDistance(char *arra, char *arrb, int asize, int bsize)
{
		int return_val, prev;
		int **matrix = malloc(sizeof(int *) * (asize + 1));
		for (int i = 0; i <= asize; i++)
		{
				*(matrix + i) = malloc(sizeof(int) * (bsize + 1));
				memset(*(matrix + i), 0, sizeof(int) * (bsize + 1));
		}
		for (int i = 1; i <= asize; i++)
				**(matrix + i) = i;
		for (int i = 1; i <= bsize; i++)
				*((*matrix) + i) = i;
		for (int i = 1; i <= asize; i++)
				for (int j = 1; j <= bsize; j++)
				{
						if (arra[i - 1] == arrb[j - 1])
								prev = 0;
						else
								prev = 1;
						*(*(matrix + i) + j) = min(*(*(matrix + i) + j) + prev,
												   *(*(matrix + i - 1) + j) + 1,
												   *(*(matrix + i) + j -1) + 1);
				}
		return(*(*(matrix + asize) + bsize));
}

int main(void)
{
		int result;
		struct timeb ts1, ts2;
		ftime(&ts1);
		char *a = "Oh my gooiidood";
		char *b = "hh mv gooioeoid";
		printf("\n\t\t%s\n\t\t%s\n\t\tCalling editDistanceRecursive:\n\t\t", a, b);
		result = editDistanceRecursive(a, b, strlen(a), strlen(b));
		ftime(&ts2);
		printf("Result: %d, Cost: %lf seconds\n\t\t", result, (double)((ts2.time - ts1.time) * 1000 + (ts2.millitm - ts1.millitm)) / 1000.0);
		printf("Calling Dynamic Programming editDistance:\n\t\t");
		ftime(&ts1);
		editDistance(a, b, strlen(a), strlen(b));
		ftime(&ts2);
		printf("result: %d, Cost: %lf seconds\n", result, (double)((ts2.time - ts1.time) * 1000 + (ts2.millitm - ts1.millitm)) / 1000.0);
		return(0);
}

File: Dynamic Programming/edit_distance.py
import time

def pr_matrix(matrix):
    for i in matrix:
        for j in i:
            print("%2d " % j, end="")
        print("")

def editDistanceRecursive(a, b):
    if len(a) == 0:
        return len(b)
    if len(b) == 0:
        return len(a)
    edit = 0 if a[-1] == b[-1] else 1
    return min(editDistanceRecursive(a[:-1], b[:-1]) + edit,
               editDistanceRecursive(a[:-1], b) + 1,
               editDistanceRecursive(a, b[:-1]) + 1)

def editDistance(a, b):
    matrix = []
    for i in range(len(a) + 1):
        matrix.append([0] * (len(b) + 1))
    for i in range(len(a) + 1):
        matrix[i][0] = i
    for i in range(len(b) + 1):
        matrix[0][i] = i
    for i in range(1, len(a) + 1):
        for j in range(1, len(b) + 1):
            horize = matrix[i][j - 1] + 1
            vert = matrix[i -1][j] + 1
            distdlg = matrix[i - 1][j - 1] if a[i - 1] == b[j - 1] else matrix[i -1][j - 1] + 1
            matrix[i][j] = min(horize, vert, distdlg)
    #pr_matrix(matrix)
    return matrix[-1][-1]


if __name__ == "__main__":
    a = "Oh my gooood"
    b = "hh mv goooid"
    print("\n\t\t%s\n\t\t%s\n\t\t" % (a, b), end="")
    t1 = time.time()
    print("editDistanceRecursive result:", editDistanceRecursive(a, b), end="")
    t2 = time.time()
    print("\tCost: %.5f" % (t2 - t1), "seconds\n\t\t", end="")
    t1 = time.time()
    print("Dynamic Programming Editdistance result: ", editDistance(a, b), end="")
    t2 = time.time()
    print("\tCost: %.5f" % (t2 - t1), "seconds\n\t\t", end="")

